src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -1720,4 +1726,200 @@ public void run() {
+        }
+        engine.refresh("test");
+        try (Engine.Searcher searcher = engine.acquireSearcher("test")) {
+            TopDocs topDocs = searcher.searcher().search(new MatchAllDocsQuery(), randomIntBetween(numDocs, numDocs + 10));
+            assertThat(topDocs.totalHits, equalTo(numDocs));
+        }
+        final MockDirectoryWrapper directory = DirectoryUtils.getLeaf(store.directory(), MockDirectoryWrapper.class);
+        if(directory != null) {
+            directory.setPreventDoubleWrite(false);
+        }
+        long currentTranslogId = translog.currentId();
+        engine.close();
+        engine = new InternalEngine(engine.config(), true);
+        assertTrue(currentTranslogId + "<" + translog.currentId(), currentTranslogId < translog.currentId());
+        assertEquals("translog ID must be incremented by 2 after initial recovery", currentTranslogId+2, translog.currentId());
+
+        try (Engine.Searcher searcher = engine.acquireSearcher("test")) {
+            TopDocs topDocs = searcher.searcher().search(new MatchAllDocsQuery(), randomIntBetween(numDocs, numDocs+10));
+            assertThat(topDocs.totalHits, equalTo(0));
+        }
+
+    }
+
+    public void testTranslogReplay() throws IOException {
+        boolean canHaveDuplicates = true;
+        boolean autoGeneratedId = true;
+        final int numDocs = randomIntBetween(1, 10);
+        for (int i = 0; i < numDocs; i++) {
+            ParsedDocument doc = testParsedDocument(Integer.toString(i), Integer.toString(i), "test", null, -1, -1, testDocument(), new BytesArray("{}"), false);
+            Engine.Create firstIndexRequest = new Engine.Create(null, newUid(Integer.toString(i)), doc, Versions.MATCH_ANY, VersionType.INTERNAL, PRIMARY, System.nanoTime(), canHaveDuplicates, autoGeneratedId);
+            engine.create(firstIndexRequest);
+            assertThat(firstIndexRequest.version(), equalTo(1l));
+        }
+        engine.refresh("test");
+        try (Engine.Searcher searcher = engine.acquireSearcher("test")) {
+            TopDocs topDocs = searcher.searcher().search(new MatchAllDocsQuery(), randomIntBetween(numDocs, numDocs + 10));
+            assertThat(topDocs.totalHits, equalTo(numDocs));
+        }
+        final MockDirectoryWrapper directory = DirectoryUtils.getLeaf(store.directory(), MockDirectoryWrapper.class);
+        if(directory != null) {
+            directory.setPreventDoubleWrite(false);
+        }
+
+        final TranslogHandler parser = (TranslogHandler) engine.config().getTranslogRecoveryPerformer();
+        long currentTranslogId = translog.currentId();
+        parser.mappingModified = randomBoolean();
+
+        engine.close();
+        engine = new InternalEngine(engine.config(), false); // we need to reuse the engine config unless the parser.mappingModified won't work
+        assertTrue(currentTranslogId + "<" + translog.currentId(), currentTranslogId < translog.currentId());
```

Comment:
we can use assertThat(translog.currentId(), isGreaterThan(currentTranslogId) and get nice messages automatically.